### ISSUE TITLE
First Release - First Issue! Whats next...

### ISSUE THREAD

**Issue by draabe** [floodlight-sports/floodlight#38]
Hello Everybody!

It's time to celebrate a new year and, a little belated, our very first release! :tada: :fireworks:

A big :balloon:**THANK YOU**:balloon: to all contributors who made this first fully functional release possible in just two months! I'm incredibly excited to see this package finally being released. It's been four years of a seemingly endless conceptualize-implement-refactor cycle for me, and I'm more than grateful for all your help in realizing this project!

At this point, we've got all the essentials ready, so the next steps will be to gradually add features and functionality to our package. With this first issue, I just want to give a quick overview of the topics (and few bugs) we've come across preparing the first release, and a comprehensive overview over the potential next steps.

Please note that I'm kind of squashing everything in to a single issue which is not the typical way to go. But instead of opening a hundred single issues, I'd like to keep this one for now as a primary feed for discussing the next release. It's really just for convenience and acknowledging the rapid development phase we are in at this stage. I will also include a more detailed long-term vision and project outline soon.

In this sense, this is not the right thread to collect long-term feature ideas (which would probably explode at this point), but rather a collection of useful functionality that can be included right away on top of the current release.

*However, feature requests are highly welcome, please just open a separate issue! Same goes for bug fixes or documentation additions/correcionts - we have issue templates ready for all of these.*

It's still a few basic processing routines we need to include before we can start jumping into really fancy analytics, but I'm optimistic we can include the first (semantic) processing routines soon! If you are interested in implementing/fixing one of the points, just leave a comment!

## Fixes

A list of known bugs and errors that should be fixed:

- [x] ``'tracab'`` pitch template is in m, but data is in cm
- [x] codes returned by DFL parser should be np.arrays, not lists
- [x] event ``'minutes'`` and ``'seconds'`` columns returned by Opta parser should be relative to segments
- [x] update python version in pyproject.toml to match latest version on runner

## Improvements

A list of things that should be improved for the existing code:

- [x] #70
- [x] Include missing tests for
  - [x] Events.column_values_in_range
  - [x] all Events' properties

## Potential Features for Upcoming Release

A list of features that are of interest for the next upcoming release. 

#### XY

- [ ] ``def permute(self, i: int, j: int)``: Permute data (i.e. switch columns) of player with xIDs *i* and *j*.
- [ ] ``def estimate_playing_direction(self, pitch)``: Estimate ``XY.direction`` property from data and pitch.
- [ ] merge columns
- [ ] missing data checks

#### Events

- [x] Spatial transforms mimicking the ones for XY objects:
  - ``def translate(self, shift: Tuple[Numeric, Numeric])``
  - ``def scale(self, factor: float, axis: int = None)``
  - ``def reflect(self, axis: int)``
  - ``def rotate(self, alpha: float)``
- [x] Slicing method mimicking the one of the``XY`` class, although based on gameclock or frameclock column.
- [ ] ``def estimate_playing_direction(self, pitch)``: Estimate ``XY.direction`` property from data and pitch.
- [x] ``def get_event_stream(self, fade: int=0) -> Code``: Generate a Code (continuous) object from a list of events, maybe makes more sense to make this a property. 

#### Code

- [x] Implement wrapper/dunders for numpy's logical indexing, i.e. to enable ``code[code == 'A']``
- [x] Find sequences, i.e., compute a list of start- and end-frames for sequences of consecutive frames with the same token:
  ``['A', 'A', 'H', 'H', 'H', 'H', 'A', 'A', 'A']`` -> ``[(0, 2, 'A'), (2, 6, 'H'), (6, 9, 'A')]``

#### Pitch

- [x] Property methods for sport-specific coordinates of pitch markings or goal positions

#### New classes

- [ ] Add core classes and basic functionality:
  - [x] Property
  - [ ] Teamsheet

#### Plotting

- [x] Basic ``plot(...)`` methods for major core objects. Plots on a given axes or creates one if not supplied. An initial kick-off could incorporate:
  - [x] pitch: Plot a (background) playing surface for a given sport @MartinWohlan 
  - [x] xy: Plot position data for a time frame, or trajectories for a time period @MartinWohlan 
  - [ ] events: Plot event data (locations) for a time period or event type (summary view)
  - [ ] code: Plot temporal summary

#### New parser

- [x] SecondSpectrum Position Data
- [x] Statsperform general format
- [ ] Opta f7/f9 feeds

  

### Utils-Module

- [x] Refactor and test ``def get_and_convert`` from Opta parser as utility function

### New Modules

New modules to be added soon and attached functionality:

- [ ] ``floodlight.transforms``: This module includes (mathematical) transformations performed on more than one object. These include spatial and temporal transformations and generally consists of functions that perform operations of the form 
  ``Tuple[data-level core objects]`` -> ``Tuple[data-level core objects]``.

  - [ ] Temporal transformations
    - [ ] Downsampling: Downsample framerate-bound core objects. This may involve data interpolation and is rather non-trivial to implement.
    - [ ] Upsampling: Upsample framerate-bound core objects. This may involve data imputation and is rather non-trivial to implement.
    - [ ] Synchronization: Given multiple objects, re-sample them cleverly to the same framerate.
  - [ ] Spatial transformations
    - [ ] Extend core object methods such as scaling, rotating, reflecting to the multi-object case that [ ] also adjusts the coordinate system coded in the ``Pitch`` object.
    - [ ] Synchronization: Given multiple objects, transform them cleverly onto the same pitch.

- [x] ``floodlight.models``: This module is intended to collect algorithms that process sports data and can be roughly described as model building on a frame-by-frame level (in contrast, e.g., to aggregated performance metric calculation). These operations can roughly be described as functions that perform operations of the form 
  ``Tuple[data-level core objects]`` -> ``Union[Code-objects, Property-Objects]``.

  An example would be one of the many space-control models that calculates, for each frame and player, the amount of space he or she controls. There are countless models we could implement, but for starters I would suggest the following three:

  - [x] Physiological Models: May include basic kinematics such as speed, acceleration, distance covered, number of sprints, speed bands, load metrics, ... . @manuba95 
  - [x] "Basic" Geometric Models: These include rather straightforward computations regarding player-to-player or player-to-ball distances, convex hull, centroids, stretch, ... .
  - [ ] Space Control Models: These models can be mostly found in football. Of the many available, I would start with the most basic Voronoi model before anything else.

## Documentation

Possible upgrades to our docs for the next release:

#### Contributing Manual

- [ ] Extend the Contributing Manual for the following subchapters:
  - [ ] How to do a Pull Request
  - [ ] How to submit an Issue
  - [ ] How to document your Code with numpy docstrings and Sphinx

#### Tutorials

- [x] Add basic tutorials (exceeding the getting started chapter) for data handling using the package

#### Compendium

- [ ] Add examples to the existing chapters for clarity:
  - [ ] Handling different Clocks
  - [ ] Which data / raw data files goes into which core objects

---

Last but not least: I really hope our joint interest in sports data can help us build a small welcoming, constructive and inspiring community around this project! So with opening the Issue section, I'd like to encourage everyone to use Issues and Pull Requests for contributions as well as general communication! :call_me_hand:

**Comment by MartinWohlan** [comment]
I would like to contribute to the project and build a plot function for the Pitch object.

**Comment by MartinWohlan** [comment]
Thanks for assigning me. Subsequently, when the Pull Request for the Pitch.plot() is finished I would like to deal with the task of plotting position data for a certain time frame or trajectories for a time period.
